sanic/websocket.py

```diff
@@ -7,23 +10,748 @@
+                assert False, "ping_id is in self.pings"
+            # Remove acknowledged pings from self.pings.
+            for ping_id in ping_ids:
+                del self.pings[ping_id]
+
+    async def keepalive_ping(self) -> None:
+        """
+        Send a Ping frame and wait for a Pong frame at regular intervals.
+        This coroutine exits when the connection terminates and one of the
+        following happens:
+        - :meth:`ping` raises :exc:`ConnectionClosed`, or
+        - :meth:`auto_close_connection` cancels :attr:`keepalive_ping_task`.
+        """
+        if self.ping_interval is None:
+            return
+
+        try:
+            while True:
+                await asyncio.sleep(self.ping_interval)
+
+                # ping() raises CancelledError if the connection is closed,
+                # when auto_close_connection() cancels self.keepalive_ping_task.
+
+                # ping() raises ConnectionClosed if the connection is lost,
+                # when connection_lost() calls abort_pings().
+
+                ping_waiter = await self.ping()
+
+                if self.ping_timeout is not None:
+                    try:
+                        await asyncio.wait_for(ping_waiter, self.ping_timeout)
+                    except asyncio.TimeoutError:
+                        error_logger.warning("Websocket timed out waiting for pong")
+                        self.fail_connection(1011)
+                        break
+        except asyncio.CancelledError:
+            raise
+        except ConnectionClosed:
+            pass
+        except BaseException:
+            error_logger.warning("Unexpected exception in keepalive ping task")
+
+    def fail_connection(self, code: int = 1006, reason: str = "") -> bool:
+        """
+        Fail the WebSocket Connection
+        This requires:
+        1. Stopping all processing of incoming data, which means cancelling
+           pausing the underlying io protocol. The close code will be 1006
+           unless a close frame was received earlier.
+        2. Sending a close frame with an appropriate code if the opening
+           handshake succeeded and the other side is likely to process it.
+        3. Closing the connection. :meth:`auto_close_connection` takes care
+           of this.
+        (The specification describes these steps in the opposite order.)
+        """
+        if self.io_proto.transport is not None:
+            # Stop new data coming in
+            # In Python Version 3.7: pause_reading is idempotent
+            # i.e. it can be called when the transport is already paused or closed.
+            self.io_proto.transport.pause_reading()
+
+            # Keeping fail_connection() synchronous guarantees it can't
+            # get stuck and simplifies the implementation of the callers.
+            # Not draining the write buffer is acceptable in this context.
+
+            # clear the send buffer
+            _ = self.connection.data_to_send()
```

Comment:
This works because `data_to_send()` always aligns with frame boundaries and you can interject a close frame anywhere. (Mentioning it because it wasn't immediately obvious to me that this was legal.)

Reply:
Yeah, I tried it a few ways, and I concluded that this is fine. On `fail_connection` we no longer care about may or not need to be send from the send buffer, but we also need to clear it completely in order to get the Close frame, then send that.